crates/gossip/src/gossip.rs

```diff
@@ -26,132 +35,634 @@ impl K2GossipFactory {
 }
 
 impl K2Gossip {
     /// Construct a new [K2Gossip] instance.
+    #[allow(clippy::too_many_arguments)]
     pub fn create(
-        space: SpaceId,
+        config: K2GossipConfig,
+        space_id: SpaceId,
+        space: DynSpace,
         peer_store: DynPeerStore,
+        peer_meta_store: DynPeerMetaStore,
         op_store: DynOpStore,
         transport: DynTransport,
-    ) -> DynGossip {
+        agent_verifier: DynVerifier,
+    ) -> K2Gossip {
         let (response_tx, mut rx) =
             tokio::sync::mpsc::channel::<GossipResponse>(1024);
         let response_task = tokio::task::spawn({
-            let space = space.clone();
+            let space_id = space_id.clone();
             let transport = transport.clone();
             async move {
                 while let Some(msg) = rx.recv().await {
-                    transport
+                    if let Err(e) = transport
                         .send_module(
                             msg.1,
-                            space.clone(),
+                            space_id.clone(),
                             MOD_NAME.to_string(),
                             msg.0,
                         )
                         .await
-                        .unwrap();
+                    {
+                        tracing::error!("could not send response: {:?}", e);
+                    };
                 }
             }
         })
         .abort_handle();
 
         let gossip = K2Gossip {
-            space: space.clone(),
-            _peer_store: peer_store,
-            _op_store: op_store,
+            config: Arc::new(config),
+            initiated_round_state: Arc::new(Mutex::new(None)),
+            accepted_round_states: Arc::new(Mutex::new(HashMap::new())),
+            space_id: space_id.clone(),
+            space: Arc::downgrade(&space),
+            peer_store,
+            peer_meta_store: Arc::new(K2PeerMetaStore::new(
+                peer_meta_store,
+                space_id.clone(),
+            )),
+            op_store,
+            agent_verifier,
             response_tx,
-            response_task,
+            _response_task: Arc::new(DropAbortHandle {
+                name: "Gossip response task".to_string(),
+                handle: response_task,
+            }),
         };
 
         transport.register_module_handler(
-            space,
+            space_id,
             MOD_NAME.to_string(),
             Arc::new(gossip.clone()),
         );
 
-        Arc::new(gossip)
+        gossip
     }
 }
 
 impl K2Gossip {
+    // TODO dead code until the initiate task is created
+    #[allow(dead_code)]
+    pub(crate) async fn initiate_gossip(
+        &self,
+        target: AgentId,
+    ) -> K2Result<bool> {
+        let state = self.initiated_round_state.clone();
+        let mut initiated_lock = state.lock().await;
+        if initiated_lock.is_some() {
+            tracing::info!("initiate_gossip: already initiated");
```

Comment:
I imagine this is not supposed to happen if everything works correctly. Should these "already doing this" messages be at `warn` level?

Reply:
End users won't be able to do anything with that though... I'd say it should go the other way and be a debug.

Reply:
Agree, it's not actionable by an end user. It's something we might track and block if somebody keeps doing it, in some fancy future version with a stronger security stance.

Changed to `debug!`